### PR TITLE
Extract GroupValues (#6969)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #6969

# Rationale for this change

Extracts the group key storage into a type-erased trait object so that it can then be specialized for different column types.

I have not been able to run the benchmarks on this yet, but I don't expect it to have a major material impact

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 1]
```diff
@@ -385,102 +492,25 @@ impl RecordBatchStream for GroupedHashAggregateStream {
...
-                        allocated,
-                    );
-                    group_idx
-                }
-            };
-            group_indices.push(group_idx);
-        }
-
-        // account for memory growth in scratch space
-        *allocated += self.scratch_space.size();
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:451]: I couldn't see a reason to keep track of the delta and not just use try_resize at the end

**Comment by alamb** [comment]
I am running benchmarks on this PR

**Comment by alamb** [comment]
My benchmark runs confirm what @tustvold  suspected that there is no major performance change for this PR

```
--------------------
Benchmark tpch_mem.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ main_base ┃ extract-group-values ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  570.35ms │             567.53ms │     no change │
│ QQuery 2     │  166.07ms │             162.08ms │     no change │
│ QQuery 3     │  174.89ms │             170.65ms │     no change │
│ QQuery 4     │  115.82ms │             113.91ms │     no change │
│ QQuery 5     │  390.04ms │             388.49ms │     no change │
│ QQuery 6     │   39.12ms │              40.99ms │     no change │
│ QQuery 7     │  956.00ms │             925.58ms │     no change │
│ QQuery 8     │  239.53ms │             249.49ms │     no change │
│ QQuery 9     │  586.23ms │             586.65ms │     no change │
│ QQuery 10    │  331.74ms │             339.71ms │     no change │
│ QQuery 11    │  166.03ms │             155.13ms │ +1.07x faster │
│ QQuery 12    │  173.20ms │             174.27ms │     no change │
│ QQuery 13    │  305.51ms │             306.92ms │     no change │
│ QQuery 14    │   48.78ms │              51.24ms │  1.05x slower │
│ QQuery 15    │   52.68ms │              53.09ms │     no change │
│ QQuery 16    │  165.92ms │             172.41ms │     no change │
│ QQuery 17    │  981.34ms │             924.09ms │ +1.06x faster │
│ QQuery 18    │ 1644.03ms │            1696.96ms │     no change │
│ QQuery 19    │  169.66ms │             168.50ms │     no change │
│ QQuery 20    │  334.96ms │             329.97ms │     no change │
│ QQuery 21    │ 1161.63ms │            1189.21ms │     no change │
│ QQuery 22    │   87.75ms │              90.87ms │     no change │
└──────────────┴───────────┴──────────────────────┴───────────────┘
```

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 2]
```diff
@@ -60,6 +60,151 @@ pub(crate) enum ExecutionState {
 
 use super::AggregateExec;
 
+/// An interning store for group keys
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:63]:
```suggestion
/// Stores group key values and their mapping to group_index
///
/// This is a trait to allow special casing certain kinds of keys 
/// like single column primitive arrays
```

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 3]
```diff
@@ -60,6 +60,151 @@ pub(crate) enum ExecutionState {
...
+    fn is_empty(&self) -> bool;
+
+    /// The number of values stored in this [`GroupValues`]
+    fn len(&self) -> usize;
+
+    /// Flushes the unique [`GroupValues`]
+    fn flush(&mut self) -> Result<Vec<ArrayRef>>;
+}
+
+/// A [`GroupValues`] making use of [`Rows`]
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:81]:
```suggestion
/// A [`GroupValues`] which stores group values using the arrow_row format [`Rows`]
```

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 4]
```diff
@@ -60,6 +60,151 @@ pub(crate) enum ExecutionState {
...
+
+    /// Logically maps group values to a group_index in
+    /// [`Self::group_values`] and in each accumulator
+    ///
+    /// Uses the raw API of hashbrown to avoid actually storing the
+    /// keys (group values) in the table
+    ///
+    /// keys: u64 hashes of the GroupValue
+    /// values: (hash, group_index)
+    map: RawTable<(u64, usize)>,
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:94]: I am surprised to see the `map` in this structure -- I was expecting only the group values.

Is your idea to store the group keys inside the map, as shown to be so effective by @yahoNanJing in https://github.com/apache/arrow-rs/pull/4524#issuecomment-1638970315 ?

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 5]
```diff
@@ -60,6 +60,151 @@ pub(crate) enum ExecutionState {
 
 use super::AggregateExec;
 
+/// An interning store for group keys
+trait GroupValues: Send {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:64]: Eventually I would love to see this in its own module `aggregates/group_values.rs` but that can be a follow on PR for sure

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 6]
```diff
@@ -60,6 +60,151 @@ pub(crate) enum ExecutionState {
...
+        // tracks to which group each of the input rows belongs
+        groups.clear();
+
+        // 1.1 Calculate the group keys for the group values
+        let batch_hashes = &mut self.hashes_buffer;
+        batch_hashes.clear();
+        batch_hashes.resize(n_rows, 0);
+        create_hashes(cols, &self.random_state, batch_hashes)?;
+
+        for (row, &hash) in batch_hashes.iter().enumerate() {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:156]: This change will likely conflict with https://github.com/apache/arrow-datafusion/pull/6932#pullrequestreview-1536741909 but I think it should be straightforward to update

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 7]
```diff
@@ -75,29 +220,29 @@ use super::AggregateExec;
...
-/// │   │  1  │───┼─┘    │            │    ││ │└─────┘ │ ││       ││ │└─────┘ │ ││
-/// │   ├─────┤   │      │            │    ││ │        │ ││       ││ │        │ ││
-/// │   │ 13  │───┼─┐    │ ┌────────┐ │    ││ │  ...   │ ││       ││ │  ...   │ ││
-/// │   └─────┘   │ └────┼▶│  "Q"   │ │    ││ │        │ ││       ││ │        │ ││
-/// └─────────────┘      │ └────────┘ │    ││ │┌─────┐ │ ││       ││ │┌─────┐ │ ││
-///                      │            │    ││ │└─────┘ │ ││       ││ │└─────┘ │ ││
-///                      └────────────┘    ││ └────────┘ ││       ││ └────────┘ ││
-///                                        │└────────────┘│       │└────────────┘│
-///                                        └──────────────┘       └──────────────┘
+///         ┌────────────┐              ┌──────────────┐       ┌──────────────┐
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:227]: If we choose to go with this approach, I can make some pictures for RowGroupValues as well

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/aggregates/row_hash.rs:451]: Related response: https://github.com/apache/arrow-datafusion/pull/6932#discussion_r1267842060

**Pull request merged**